proto-loader-gen-types Typename templates

install opened the pull request:
- Allow for customizing the naming pattern for both restricted and permissive types

Enums and filenames are left untouched

murgatroid99 commented:
Can you expand some more on what your use case is for this, and also why you excluded enums?

install commented:
@murgatroid99 
I've found that for the majority of my use cases I'm consuming the `__Output` types, so it'd be nice if the consumer can determine the naming format for inputs vs outputs.

The enum is a special case because it's both a type and a value. Parent messages seem to reference the enum directly.

```ts
export interface TestMessage {
  'enum'?: (test_v1_Enum | keyof typeof test_v1_Enum);
}

export interface TestMessage__Output {
  'enum': (keyof typeof test_v1_Enum);
}

```

It might be worth treating enums like messages in that we create both an input and an output version, and the input/output parent message would point to the corresponding one. I wanted to leave that for a different PR for now.

install commented:
Hey @murgatroid99 have you had a chance to look at this? I'm open to any feedback and suggestions :) thanks

murgatroid99 commented on packages/proto-loader/bin/proto-loader-gen-types.ts:
```diff
@@ -114,15 +127,16 @@ function getTypeInterfaceName(type: Protobuf.Type | Protobuf.Enum | Protobuf.Ser
   if (dependency.parent instanceof Protobuf.Type) {
     if (dependency instanceof Protobuf.Type) {
-      importedTypes = `${typeInterfaceName}, ${typeInterfaceName}__Output`;
+      importedTypes = `${typeInterfaceName}, ${outputName(typeInterfaceName)}`;
```
Should the first instance of `typeInterfaceName` on this line be `inputName(typeInterfaceName)`?

install replied:
Good catch. Updating the golden tests found a couple other places that are now covered

murgatroid99 commented on packages/proto-loader/bin/proto-loader-gen-types.ts:
```diff
@@ -805,7 +826,11 @@ async function runScript() {
       }
-    }).alias({
+    })
+    .string(['inputTemplate', 'outputTemplate'])
```
Please move this into the existing `string` call earlier, and group the new `default` calls with the others.

murgatroid99 reviewed (changes requested):
I am not confident that `inputName` is called everywhere it is needed. Since the default behavior is to leave the string unchanged, the current tests will not catch if it is omitted anywhere. Can you add a test with a different `--inputTemplate` option?

Please also create a PR to add information about these new options to [this file](https://github.com/grpc/proposal/blob/master/L70-node-proto-loader-type-generator.md).

install commented:
@murgatroid99 Thanks for the feedback! I've addressed your comments and updated the golden tests to use the new flags. 
Here's the corresponding proposal PR: https://github.com/grpc/proposal/pull/326

snarky-puppy commented:
FYI - prefixing interfaces with "I" conflicts with Hungarian notation commonly misused in other parts of the TS ecosystem. Suggest prefixing with "In"/"Out" rather than "I/O"

install commented:
@murgatroid99 How can I help push this forward?

murgatroid99 commented:
Sorry about the delay here.

murgatroid99 commented:
This is out in version 0.7.3
